lib/state_machine.js

```diff
@@ -27,99 +27,196 @@ limitations under the License.
-  // A state machine implements a specific interface and has associated states
-  // that also implement this interface. The state machine's impl forwards these
-  // methods onto the current state which might cause a transition to a new
-  // state. Transitions are effectively instantaneous, while a state might be
-  // current for some time.
-  var StateMachine = function(debug, initialState) {
-    // Just for logging purposes.
-    this.debug_ = debug;
-
-    this.current_ = new State('_initial');
-
-    // A list of predicate(state), defer pairs. When we enter a state, we run
-    // the predicates. If any return true, we resolve the defer & remove the
-    // pair from the list.s
-    this.predicates_ = [];
-
-    // An arbitrary struct of stuff that is 'global' w.r.t this state machine
-    // instance. Subclasses can set this, and this object will be shared with
-    // every 'State' instance before 'enter' is called.
-    this.context_ = {};
-
-    // We _really_ start in the initial state.
-    this.enterNewState_(initialState);
-  };
-  StateMachine.prototype.enterNewState_ = function(state) {
-    if (state === this.current_) {
-      // No work to do, as we're already in that state!
-      this.debug_('Self-edge in state machine!', state.name_);
-      return Promise.reject('self-edge');
-    }
+  /**
+   * A state interface. All states must inherit from this interface.
+   * @interface
+   */
+  class State {
+    /**
+     * Return the name of this state.
+     * @return {string}
+     */
+    getName() { return this.constructor.name; }
+  
+    /**
+     * Called by the machine when we transition to this state.
+     * @abstract
+     * @param {!function(!State)} transition A function that, if called, will
+     *    drive the state machine to the specified state.
+     * @param {*} context A context object shared by all states in the same
+     *    machine.
+     */
+    enter(transition, context) {}
+  
+    /**
+     * Called by the machine when we transition away from this state.
+     * @abstract
+     */
+    exit() {}
+  }
 
-    this.debug_('Transition from', this.current_.name_, 'to', state.name_);
-    this.current_ = state;
-    // Add context to this state.
-    state.setContext(this.context_);
-    try {
-      state.enter_();
-    } catch (e) {
-      this.debug_('While entering', state.name_);
-      this.debug_(e.message);
-      this.debug_(e.stack);
+  /**
+   * A state machine.
+   * @template STATETYPE
+   */
+  class StateMachine {
+    constructor(initialState, debug = undefined) {
+      // We don't transition to the initial state; we are simply in that state to
+      // begin with.
+      /** @type {!STATETYPE} */
+      this.state = initialState;
+    
+      /** @private @const */
+      this.debug_ = debug;
+    
+      /** @private {?STATETYPE} */
+      this.nextExternalState_ = null;
+  
+      /**
+       * Resolver for the transition promise.
+       * @private {!function(!STATETYPE)}
```

Comment:
This is google-standard JS style. These are type annotations that would tell the jscompiler what to do. I wrote them out of habit, and we can choose to keep them as documentation if we want.